packages/gatsby/src/datastore/in-memory/indexing.ts

```diff
@@ -21,6 +24,88 @@ export type FilterOp =  // TODO: merge with DbComparator ?
 // Note: `undefined` is an encoding for a property that does not exist
 
 export type FilterCacheKey = string
+type GatsbyNodeID = string
+
+export interface IGatsbyNodePartial {
+  isGatsbyNodePartial: boolean
+  id: GatsbyNodeID
+  internal: {
+    counter: number
+  }
+  indexFields: Set<string>
+  [k: string]: any
+}
+
+const nodeIdToIdentifierMap = new Map<
+  GatsbyNodeID,
+  WeakRef<IGatsbyNodePartial>
+>()
+
+/**
+ * Grabs an instance of IGatsbyNodePartial for the given node.
+ * This accepts an IGatsbyNode or IGatsbyNodePartial as input, which allows
+ * us to conditionally store index fields on the partial if we encounter
+ * one that hasn't been stored on the partial yet.
+ */
+export const getGatsbyNodePartial = (
+  node: IGatsbyNode | IGatsbyNodePartial,
+  indexFields: Array<string>,
+  resolvedFields: any
```

Comment:
We should try to not use any, wouldn't the below type work?
```suggestion
  resolvedFields: Record<string, unknown>
```

Reply:
Good call, will update. Forgot to fix that in this file!